_source/_posts/2022-06-01-jhipster-digital-ocean.md

```diff
@@ -0,0 +1,411 @@
+
+## Set up a microservices architecture for Kubernetes
+
+Before working on the application, you need to install JHipster. The classical way of working with JHipster is to do a local installation with NPM.
+
+```bash
+npm install -g generator-jhipster@7
+```
+
+If you'd rather use Yarn or Docker, follow the instructions at [jhipster.tech](https://www.jhipster.tech/installation/#local-installation-with-npm-recommended-for-normal-users).
+
+For this test, start from the `reactive-jhipster` example in the `java-microservices-examples` repository on [GitHub](https://github.com/oktadev/java-microservices-examples). The example is a JHipster reactive microservices architecture with Spring Cloud Gateway and Spring WebFlux, Vue as the client framework, and Gradle as the build tool. You can read about how it was built in [Reactive Java Microservices with Spring Boot and JHipster](/blog/2021/01/20/reactive-java-microservices).
+
+```bash
+git clone https://github.com/oktadev/java-microservices-examples.git
+cd java-microservices-examples/reactive-jhipster
+```
+If you inspect the project folder, you will find sub-folders for the `gateway` service, which will act as the front-end application, and a gateway to the `store` and `blog` microservices, which also have their subfolders. A `docker-compose` sub-folder contains the service definitions for running the application containers.
+
+The next step is to generate the Kubernetes deployment descriptors. In the project root folder, create a `k8s` directory and run the `k8s` JHipster sub-generator:
+
+```bash
+mkdir k8s
+cd k8s
+jhipster k8s
+```
+
+Choose the following options when prompted:
+
+- Type of application: **Microservice application**
+- Root directory: **../**
+- Which applications? (select all)
+- Set up monitoring? **No**
+- Which applications with clustered databases? select **store**
+- Admin password for JHipster Registry: (generate one)
+- Kubernetes namespace: **demo**
+- Docker repository name: (your docker hub username)
+- Command to push Docker image: `docker push`
+- Enable Istio? **No**
+- Kubernetes service type? **LoadBalancer**
+- Use dynamic storage provisioning? **Yes**
+- Use a specific storage class? (leave empty)
+
+**NOTE**: You can leave the Docker repository name blank for running Kubernetes locally, but a repository will be required for the cloud deployment, so go ahead and create a [Docker Hub](https://hub.docker.com/) personal account, and the image pull configuration will be ready for both local and cloud deployments.
+
+Build the `gateway`, `store` and `blog` services container images with Jib. For example, for the `gateway` service:
+
+```bash
+cd ../gateway
+./gradlew bootJar -Pprod jib -Djib.to.image=<docker-repo-name>/gateway
```

Comment:
I get an error here on my M1 MacBook Pro:

```
npm ERR! path /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/puppeteer
npm ERR! command failed
npm ERR! command sh -c node install.js
npm ERR! The chromium binary is not available for arm64.
npm ERR! If you are on Ubuntu, you can install with:
npm ERR!
npm ERR!  sudo apt install chromium
npm ERR!
npm ERR!
npm ERR!  sudo apt install chromium-browser
npm ERR!
npm ERR! /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserFetcher.js:115
npm ERR!                     throw new Error();
npm ERR!                     ^
npm ERR!
npm ERR! Error
npm ERR!     at /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserFetcher.js:115:27
npm ERR!     at FSReqCallback.oncomplete (node:fs:198:21)

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/mraible/.npm/_logs/2022-05-13T18_35_55_087Z-debug-0.log
```

I tried `rm -rf package-lock.json node_modules`, but it didn't help. This looks related: `https://github.com/puppeteer/puppeteer/issues/7740`

Reply:
I tried nuking the whole thing with `jhipster jdl reactive-ms.jdl --force` in the parent directory. It seems to update everything but also fails.

```
$ ./gradlew bootJar -Pprod jib -Djib.to.image=mraible/gateway

> Task :npm_install

up to date, audited 2029 packages in 2s

185 packages are looking for funding
  run `npm fund` for details

8 vulnerabilities (5 moderate, 3 high)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.

> Task :webapp FAILED

> gateway@0.0.0 webapp:prod
> npm run clean-www && npm run webapp:build:prod --


> gateway@0.0.0 clean-www
> rimraf build/resources/main/static/app/{src,build/}


> gateway@0.0.0 webapp:build:prod
> npm run webpack -- --mode production --env stats=minimal


> gateway@0.0.0 webpack
> webpack --config webpack/webpack.common.js "--mode" "production" "--env" "stats=minimal"

[webpack-cli] Error: Cannot find module 'acorn'
Require stack:
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/acorn-import-assertions/lib/index.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/javascript/JavascriptParser.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/javascript/JavascriptModulesPlugin.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/WebpackOptionsApply.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/webpack.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/index.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack-cli/lib/webpack-cli.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack-cli/lib/bootstrap.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack-cli/bin/cli.js
- /Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/bin/webpack.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/acorn-import-assertions/lib/index.js:8:38)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/javascript/JavascriptParser.js:9:30)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/acorn-import-assertions/lib/index.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/javascript/JavascriptParser.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/javascript/JavascriptModulesPlugin.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/WebpackOptionsApply.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/webpack.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/lib/index.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack-cli/lib/webpack-cli.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack-cli/lib/bootstrap.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack-cli/bin/cli.js',
    '/Users/mraible/Downloads/java-microservices-examples/reactive-jhipster/gateway/node_modules/webpack/bin/webpack.js'
  ]
}

FAILURE: Build failed with an exception.
```

Reply:
I tried `rm -rf blog docker-compose gateway k8s store` to start over. One thought I had is you could change the k8s section to use JDL. I tried appending this to the JDL.

```
deployment {
  deploymentType kubernetes
  appsFolders [gateway, blog, store]
  clusteredDbApps [store]
  serviceDiscoveryType eureka
  dockerRepositoryName "mraible"
}
```

When I tried to build the gateway, it fails:

```> Task :jib FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jib'.
> com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: 
Build image failed, perhaps you should make sure your credentials for 
'registry-1.docker.io/mraible/gateway' are set up correctly. 
See https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-unauthorized for help
```

I tried `docker login`, but it results in:

```
Error response from daemon: Get "https://registry-1.docker.io/v2/": unauthorized: please use personal access token to login
```

Reply:
It looks like I forgot to include the docker authentication step. Will look at this.

Reply:
I tried disabling my 2FA on my Docker account and then `docker login` worked. It also prints the following message.

```
Logging in with your password grants your terminal complete access to your account.
For better security, log in with a limited-privilege personal access token. Learn more at https://docs.docker.com/go/access-tokens/
```

I went to https://docs.docker.com/docker-hub/access-tokens/ and created an access token, then re-enabled 2FA. Now the command with `-Djib.to.image=mraible/gateway` works! We might want to mention this.

Reply:
I added a comment about `docker login` and 2FA

Reply:
FWIW, I had the JDL wrong. That's why I didn't see anything running with k9s. I was missing the `demo` namespace.

```
deployment {
  deploymentType k8s
  appsFolders [gateway, blog, store]
  clusteredDbApps [store]
  kubernetesNamespace demo
  serviceDiscoveryType eureka
  dockerRepositoryName "mraible"
}
```

Reply:
The JDL is the same as what's in the repo + the code above. However, I don't think you need to use the JDL unless you want to re-generate everything from scratch, which is probably not necessary.